Submission Export for Text/Modeling/FileUpload Exercises

schultek opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
<!-- - [ ] Client: I added multiple integration tests (Jest) related to the features -->
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Resolves #1635 

### Description
<!-- Describe your changes in detail -->

- Added a 'Export Submissions' Button for Text & Modeling & File Upload Exercises
- Added an export dialog for the above button
- Added api endpoints for each of the exercise types to generate a zip file with the exercises submissions

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to an Text, Modeling or File Upload Exercise (Course or Exam)
4. Go to the Scores Page
5. Click on 'Export Submissions' and try different options in the dialog.

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
<img width="643" alt="Bildschirmfoto 2020-07-08 um 20 22 23" src="https://user-images.githubusercontent.com/13920539/86956022-bf5c4700-c158-11ea-9c55-82c624772e22.png">
<img width="522" alt="Bildschirmfoto 2020-07-08 um 20 23 01" src="https://user-images.githubusercontent.com/13920539/86956077-d438da80-c158-11ea-8e3d-8f184138699b.png">

stefanwaldhauser commented:
![image](https://user-images.githubusercontent.com/29718932/87047122-562d1000-c1fa-11ea-95f4-d48e5a1b78a9.png)
There seem to be some missing translation strings for some errors

stefanwaldhauser commented:
I am a little confused by one export 
![image](https://user-images.githubusercontent.com/29718932/87047756-2f230e00-c1fb-11ea-9318-48c95ba81253.png)
The is only one entry in the table
but when i click on export submissions
![image](https://user-images.githubusercontent.com/29718932/87047798-3cd89380-c1fb-11ea-9a53-6e0ee1be6e63.png)

![image](https://user-images.githubusercontent.com/29718932/87047838-4cf07300-c1fb-11ea-9278-a84ebe8b1ab7.png)

stefanwaldhauser commented:
I am not able to import exported modeling exercises via apollon. Nothing is shown using import of https://apollon.ase.in.tum.de/

schultek commented:
> ![image](https://user-images.githubusercontent.com/29718932/87047122-562d1000-c1fa-11ea-95f4-d48e5a1b78a9.png)
> There seem to be some missing translation strings for some errors

Can you help me with this? I've adjusted the code from the programming exercise export.
I'm using the following on the server:
```
return ResponseEntity.badRequest()
                        .headers(HeaderUtil.createFailureAlert(applicationName, false, ENTITY_NAME, "nosubmissions", "No existing user was specified or no submission exists."))
                        .body(null);
```
And added the key "nosubmissions" to the errors.json translation files.
What am I doing wrong here?

schultek commented:
> I am a little confused by one export
> ![image](https://user-images.githubusercontent.com/29718932/87047756-2f230e00-c1fb-11ea-9318-48c95ba81253.png)
> The is only one entry in the table
> but when i click on export submissions
> ![image](https://user-images.githubusercontent.com/29718932/87047798-3cd89380-c1fb-11ea-9a53-6e0ee1be6e63.png)
> 
> Where is this second file coming from?

What do you have in participations? I think the scores page only shows assessed submissions but the export includes all.

stefanwaldhauser commented:
> > I am a little confused by one export
> > ![image](https://user-images.githubusercontent.com/29718932/87047756-2f230e00-c1fb-11ea-9318-48c95ba81253.png)
> > The is only one entry in the table
> > but when i click on export submissions
> > ![image](https://user-images.githubusercontent.com/29718932/87047798-3cd89380-c1fb-11ea-9a53-6e0ee1be6e63.png)
> > Where is this second file coming from?
> 
> What do you have in participations? I think the scores page only shows assessed submissions but the export includes all.

Ah ok I get it! If this is expected behaviour its fine

schultek commented:
> I am not able to import exported modeling exercises via apollon. Nothing is shown using import functionality of https://apollon.ase.in.tum.de/

I don't know what this is about. I just take the model string from the database and write it to a json file.
Can someone help?

TobiasPr commented:
> > I am not able to import exported modeling exercises via apollon. Nothing is shown using import functionality of https://apollon.ase.in.tum.de/
> 
> I don't know what this is about. I just take the model string from the database and write it to a json file.
> Can someone help?

The apollon version https://apollon.ase.in.tum.de/ is using a different data model to display diagrams (basically an extension), because we added some functionality which is not needed in the core library. Is it a requirement to import the diagrams to the standalone version? (edited)

alexmardale commented:
I can confirm @stefanwaldhauser 's findings.

But otherwise the export functionality worked as expected for Text and Modeling Exam exercises.

kloessst commented on src/main/webapp/app/exercises/shared/submission-export/submission-export-dialog.component.html:
```diff
@@ -0,0 +1,69 @@
+        ></textarea>
+
+        <!--
+        Only show download all checkbox for instructors & admins.
```
Should also check this on the server. if submissionExportOptions.exportAllParticipants is true but the user is just TA return forbidden()

kloessst commented on src/main/webapp/app/exercises/shared/submission-export/submission-export-dialog.component.html:
```diff
@@ -0,0 +1,69 @@
+        </button>
+        <button
+            type="submit"
+            [disabled]="!submissionExportOptions.participantIdentifierList && !submissionExportOptions.exportAllParticipants && !exportInProgress"
```
```suggestion
            [disabled]="(!submissionExportOptions.participantIdentifierList && !submissionExportOptions.exportAllParticipants) || exportInProgress"
```

kloessst commented on src/main/webapp/i18n/de/instructorDashboard.json:
```diff
@@ -34,6 +34,14 @@
+        "exportSubmissions": {
+            "title": "Abgaben exportieren",
+            "downloadAllStudents": "Oder lade die Abgaben aller Studenten herunter",
+            "question": "Tragen sie hier alle Kennungen der Studenten ein, die heruntergeladen werden sollen getrennt durch ein Komma (z.B. ga87fix,ga63dut)",
```
```suggestion
            "question": "Tragen sie hier alle Kennungen der Studenten getrennt durch ein Komma ein, deren Abgaben heruntergeladen werden sollen (z.B. ga87fix,ga63dut)",
```

Alexanderkrp commented:
Worked for me as well! 
👍

sleiss commented on src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionExportService.java:
```diff
@@ -0,0 +1,53 @@
+
+        Path submissionPath = Path.of(filePath, apiFilePathParts[apiFilePathParts.length - 1]);
+
+        if (!submissionPath.toFile().exists()) // throw if submission file does not exist
```
Please use `{` and `}`.

sleiss commented on src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionExportService.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    protected void saveSubmissionToFile(Exercise exercise, Submission submission, File file) throws IOException {
+        if (((ModelingSubmission) submission).getModel() == null) {
+            if (!file.exists())
```
Please use `{` and `}`.

sleiss commented on src/main/java/de/tum/in/www1/artemis/service/SubmissionExportService.java:
```diff
@@ -0,0 +1,242 @@
+
+        }).filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList());
+
+        if (submissionFilePaths.isEmpty())
```
Please use `{` and `}`.

sleiss commented on src/main/java/de/tum/in/www1/artemis/service/TextSubmissionExportService.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    protected void saveSubmissionToFile(Exercise exercise, Submission submission, File file) throws IOException {
+        if (((TextSubmission) submission).getText() == null) {
+            if (!file.exists())
```
Please use `{` and `}`.

sleiss commented on src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java:
```diff
@@ -274,4 +284,51 @@ else if (!authCheckService.isAtLeastTeachingAssistantForExercise(fileUploadExerc
+        }
+
+        // ta's are not allowed to download all participations
+        if (submissionExportOptions.isExportAllParticipants() && authCheckService.isTeachingAssistantInCourse(fileUploadExercise.getCourseViaExerciseGroupOrCourseMember(), null)) {
```
This check caused some problems for me during testing: If you are both instructor/admin AND tutor, you cannot download all submissions.
I do not know if this is a real scenario in normal courses, but in test courses, this can happen.
Maybe you should change the check to something like `submissionExportOptions.isExportAllParticipants() && !authCheckService.isAtLeastInstructorInCourse`.

Also, no error message was shown to me, so I had to check the source code to identify the issue. Maybe you can display error messages in a follow up or so.

sleiss commented on src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:
```diff
@@ -340,4 +351,50 @@ public ModelingExerciseResource(ModelingExerciseRepository modelingExerciseRepos
+        }
+
+        // ta's are not allowed to download all participations
+        if (submissionExportOptions.isExportAllParticipants() && authCheckService.isTeachingAssistantInCourse(modelingExercise.getCourseViaExerciseGroupOrCourseMember(), null)) {
```
See above.

alexmardale commented:
Worked as expected!
